polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h

```diff
@@ -15,10 +15,58 @@
 
 #include "mlir/Analysis/DataFlow/DenseAnalysis.h"
 #include "mlir/Dialect/Polygeist/Utils/AliasUtils.h"
+#include <set>
+#include <variant>
 
 namespace mlir {
 namespace polygeist {
 
+/// Represents a definition that is unknown (this is a singleton).
+class InitialDefinition {
+  friend raw_ostream &operator<<(raw_ostream &, const InitialDefinition &);
+
+public:
+  InitialDefinition(InitialDefinition &) = delete;
+  bool operator=(const InitialDefinition &) = delete;
+  static InitialDefinition *getInstance();
+
+private:
+  InitialDefinition() = default;
+  static InitialDefinition *singleton;
+};
+
+/// Represents either an operation that modifies a memory resource, or the
+/// initial definition.
+class Definition {
+  friend raw_ostream &operator<<(raw_ostream &, const Definition &);
+
+public:
+  Definition(Operation *op) : def(op) {}
+  Definition() : def(InitialDefinition::getInstance()) {}
+
+  bool operator==(const Definition &other) const;
+  bool operator!=(const Definition &other) const { return !(*this == other); }
+  bool operator<(const Definition &other) const;
+
+  bool isOperation() const { return std::holds_alternative<Operation *>(def); }
+  bool isInitialDefinition() const {
+    return std::holds_alternative<InitialDefinition *>(def);
+  }
+
+  Operation *getOperation() const {
+    assert(isOperation() && "expecting operation");
+    return std::get<Operation *>(def);
+  }
+
+  InitialDefinition *getInitialDefinition() const {
+    assert(isInitialDefinition() && "expecting initial definition");
+    return std::get<InitialDefinition *>(def);
+  }
+
+private:
+  std::variant<Operation *, InitialDefinition *> def;
+};
```

Comment:
This makes more sense if we avoid the singleton pattern above.

> I was not able to modify the singleton class (see comment above), however I changed the implementation of  `InitialDefinition` to avoid the memory leak. 

Also, why would we want to get the initial definition?
> The initial definition is useful in cases where a memref argument is not updated an all control-flow paths before a use. For example:

```
func.func @test(%val: i32, %cond: i1, %arg1: memref<i32>, %arg2: memref<i32>) {
  scf.if %cond {
    memref.store %val, %arg1[] : memref<i32> // <store1>
    scf.yield
  }
  %load = memref.load %arg1[] : memref<i32> // load
```
Here the reaching definition set for `arg1` (at the load operation) need to be {<initial> <store1>} because the branch may not be taken.

Reply:
Trying to understand your latest response. Are you saying we do not need to have a definition that represents the initial state ? That is not the case because we need to know whether a store "kills" the initial definition. 

This store operation in the following code kills the initial definition so the we get {store1} as the definition reaching the load:
```
func.func @test(%val: i32, %cond: i1, %arg1: memref<i32>, %arg2: memref<i32>) {
    memref.store %val, %arg1[] : memref<i32> // <store1>
  %load = memref.load %arg1[] : memref<i32> // load
```
The storein this [comment ](https://github.com/intel/llvm/pull/9319#discussion_r1189135531) on the other hand does not kill the initial definition.

Reply:
Yes, I mean the function `InitialDefinition *getInitialDefinition()` might be not needed if `InitialDefinition` is a singleton.

Reply:
I settled on creating a singleton class that uses a shared_ptr to avoid memory leaks, This member function is still needed.

Reply:
I see the issue here. This is not a singleton, but more as a placeholder type. We don't care if instances are created, move, etc., as the type will be empty and cost will be 0.

We should define the type as:
```cpp
class InitialDefinition {};
```